funnel/models/mixins.py

```diff
@@ -0,0 +1,153 @@
+# -*- coding: utf-8 -*-
+
+import urllib.parse
+
+import dateutil.parser
+import requests
+
+from coaster.utils import parse_duration
+
+from .. import app, redis_store
+from . import db
+
+
+class VideoMixin:
+    video_id = db.Column(db.UnicodeText, nullable=True)
+    video_source = db.Column(db.UnicodeText, nullable=True)
+
+    _video_duration = 0
+    _video_uploaded_at = None
+
+    @property
+    def video(self):
+        data = None
+        if self.video_source and self.video_id:
+            data = redis_store.hgetall(self.cache_key)
+            if not data:
+                data = {
+                    'source': self.video_source,
+                    'id': self.video_id,
+                    'url': self.video_url,
+                }
+                if self.video_source == 'youtube':
+                    youtube_video = requests.get(
+                        'https://www.googleapis.com/youtube/v3/videos?part=snippet,contentDetails&id={video_id}&key={api_key}'.format(
+                            video_id=self.video_id,
+                            api_key=app.config['YOUTUBE_API_KEY'],
+                        )
+                    ).json()
+                    if youtube_video is None or len(youtube_video['items']) == 0:
+                        raise Exception(
```

Comment:
Create a `VideoException`. You may also want to trap other errors from `requests` as a timeout (etc) could blow up in the UI.